website/versioned_docs/version-v2.0_alpha/developers-references/smart-contracts/MACI.md

```diff
@@ -0,0 +1,145 @@
+---
+title: MACI Smart Contract
+description: MACI main smart contract
+sidebar_label: MACI
+sidebar_position: 1
+---
+
+`MACI.sol` is the core contract of the project, as it provides the base layer for user signups and Polls to be created.
+
+The constructor shown below accepts several arguments:
+
+- `PollFactory` address
+- `MessageProcessorFactory` address
+- `TallyFactory` address
+- `SignUpGatekeeper` address
+- `InitialVoiceCreditProxy` address
+- The depth of the state tree
+
+```javascript
+constructor(
+    IPollFactory _pollFactory,
+    IMessageProcessorFactory _messageProcessorFactory,
+    ITallyFactory _tallyFactory,
+    SignUpGatekeeper _signUpGatekeeper,
+    InitialVoiceCreditProxy _initialVoiceCreditProxy,
+    uint8 _stateTreeDepth
+) payable {
+    // initialize and insert the blank leaf
+    InternalLazyIMT._init(lazyIMTData, _stateTreeDepth);
+    InternalLazyIMT._insert(lazyIMTData, BLANK_STATE_LEAF_HASH);
+
+    pollFactory = _pollFactory;
+    messageProcessorFactory = _messageProcessorFactory;
+    tallyFactory = _tallyFactory;
+    signUpGatekeeper = _signUpGatekeeper;
+    initialVoiceCreditProxy = _initialVoiceCreditProxy;
+    stateTreeDepth = _stateTreeDepth;
+
+    // Verify linked poseidon libraries
+    if (hash2([uint256(1), uint256(1)]) == 0) revert PoseidonHashLibrariesNotLinked();
+}
+
+```
+
+Upon deployment, the contract will initialize the state tree, and insert the blank leaf hash.
+
+After this, all of the parameters will be stored to state, and then the contract will perform a simple sanity check to ensure that the Poseidon hash libraries were linked successfully.
+
+Next, we have the `signUp` function, which allows users to `signUp`, as long as they pass the conditions set in the `SignUpGatekeeper` contract. This contract can use any mean necessary to gatekeep access to MACI's polls. For instance, only wallets with a specific ERC721 token can be allowed to sign up.
```

Comment:
thinking about we could add subtitle to each function, to let other page easier refer to certain function of a contract 
(e.g. /docs/developers-references/smart-contracts/MACI#signup)